### PR TITLE
Upgrade zmq

### PR DESCRIPTION
**Pull request by Sasaank**
@CJ-Wright 
extends https://github.com/NSLS-II/bluesky/pull/1064

### CODE REVIEW

**Comment by danielballan** [comment]
Thanks @Sasaank! In #1064, we had discussed removing the default prefix (hostname, pid, `id(RE)`) in favor of this new, user-configurable one. That would, of course, involve an API break, but AFAIK this aspect of the API is totally unused, and I think this simplification would be a worthwhile improvement.

Are you able to add a commit implementing that change? Once that's in place, I believe this will be ready for final review.

**Comment by Sasaank** [comment]
Hi @danielballan  I will get on that

File: bluesky/callbacks/zmq.py [hunk 1]
```diff
@@ -35,18 +36,17 @@ class Publisher:
     >>> RE = RunEngine({})
     >>> publisher = Publisher('localhost:5567', RE=RE)
     """
-    def __init__(self, address, *, RE=None, zmq=None, serializer=pickle.dumps):
+    def __init__(self, address, *, RE=None, zmq=None, prefix='',serializer=pickle.dumps):
+        if prefix is None:
```
Comment (CJ-Wright) [inline bluesky/callbacks/zmq.py]: I don't think this line is needed.

File: bluesky/callbacks/zmq.py [hunk 2]
```diff
@@ -249,12 +243,10 @@ def is_our_message(_hostname, _pid, _RE_id):
     def _poll(self):
         while True:
             message = yield from self._socket.recv()
-            hostname, pid, RE_id, name, doc = message.split(b' ', 4)
-            hostname = hostname.decode()
-            pid = int(pid)
-            RE_id = int(RE_id)
+            prefix, hostname, pid, RE_id, name, doc = message.split(b'\x00', 2)
```
Comment (CJ-Wright) [inline bluesky/callbacks/zmq.py]: This needs to have `hostname`, `pid`, and `RE_id` removed.

File: bluesky/callbacks/zmq.py [hunk 3]
```diff
@@ -57,8 +55,7 @@ def __init__(self, address, *, RE=None, zmq=None, serializer=pickle.dumps):
     def __call__(self, name, doc):
         doc = copy.deepcopy(doc)
         apply_to_dict_recursively(doc, sanitize_np)
-        message = bytes(self._prefix)  # making a copy
-        message += b' '.join([name.encode(), self._serializer(doc)])
+        message = b'\x00'.join([bytes(self._prefix),name.encode(), self._serializer(doc)])
```
Comment (CJ-Wright) [inline bluesky/callbacks/zmq.py]: this needs a space between `,` and `name`

File: bluesky/callbacks/zmq.py [hunk 4]
```diff
@@ -35,18 +36,15 @@ class Publisher:
     >>> RE = RunEngine({})
     >>> publisher = Publisher('localhost:5567', RE=RE)
     """
-    def __init__(self, address, *, RE=None, zmq=None, serializer=pickle.dumps):
+    def __init__(self, address, *, RE=None, zmq=None, prefix='',serializer=pickle.dumps):
```
Comment (CJ-Wright) [inline bluesky/callbacks/zmq.py]: PEP8

File: bluesky/tests/test_zmq.py [hunk 5]
```diff
@@ -118,6 +118,72 @@ def delayed_sigint(delay):
 
     repr(d)
 
+def test_zmq_prefix():
```
Comment (CJ-Wright) [inline bluesky/tests/test_zmq.py]: This needs the `hw` and `RE` test parameters.

**Comment by CJ-Wright** [comment]
We are stuck, it seems that test_zmq_components is doing something funky. I'm not certain what, maybe a process isn't getting killed?

**Comment by danielballan** [comment]
See https://github.com/NSLS-II/bluesky/pull/1109

**Pull request closed without merging**